compiler/modules/types/xml.bal

```diff
@@ -0,0 +1,197 @@
+// Implementation specific to basic type xml.
+
+public type XmlSubtype readonly & record {|
+    // Each xml singleton and the empty sequence is considered a primitive.
+    int primitives;
+    Bdd sequence;
+|};
+
+// Each bit represents a XML singleton subtype or a empty sequence.
+// `primitives` fields are composed of allowed values from this singleton set.
+// Atom of the `Bdd` represented by `sequence` field is also composed of these to indicate 
+// non-empty sequence of allowed singletons.
+const int XML_PRIMITIVE_NEVER      = 1;
+const int XML_PRIMITIVE_TEXT       = 1 << 1;
+const int XML_PRIMITIVE_ELEMENT_RO = 1 << 2;
+const int XML_PRIMITIVE_PI_RO      = 1 << 3;
+const int XML_PRIMITIVE_COMMENT_RO = 1 << 4;
+const int XML_PRIMITIVE_ELEMENT_RW = 1 << 5;
+const int XML_PRIMITIVE_PI_RW      = 1 << 6;
+const int XML_PRIMITIVE_COMMENT_RW = 1 << 7;
+
+const int XML_PRIMITIVE_RO_SINGLETONS = XML_PRIMITIVE_TEXT | XML_PRIMITIVE_ELEMENT_RO | XML_PRIMITIVE_PI_RO | XML_PRIMITIVE_COMMENT_RO;
+const int XML_PRIMITIVE_RO_MASK = XML_PRIMITIVE_NEVER | XML_PRIMITIVE_RO_SINGLETONS;
+const int XML_PRIMITIVE_RW_MASK = XML_PRIMITIVE_ELEMENT_RW | XML_PRIMITIVE_PI_RW | XML_PRIMITIVE_COMMENT_RW;
+const int XML_RW_ATOM = XML_PRIMITIVE_RO_SINGLETONS | XML_PRIMITIVE_RW_MASK;
```

Comment:
This name seems a bit inconsistent with the others. I would call this `XML_PRIMITIVE_SINGLETON` and rename `XML_PRIMITIVE_RO_SINGLETONS` to `XML_PRIMITIVE_RO_SINGLETON`.